fix(email): Decode subject, support for encodings other than utf-8 in IMAP folder

IgorA100 opened the pull request:
This PR holds four fixes in the context of fetching emails. Receiving emails has been refactored from v13 to v14, and by doing so, some issues have been introduced.

1. Casting a string to int: the UIDs are integers but have been inconsistent processed as a string as well as int before, causing issues when doing comparisons.
2. When there is an issue with recognizing the encoding of an email subject, this causes errors and breaks the import of new emails for an email account. This fix closes https://github.com/frappe/frappe/issues/20808 and closes https://github.com/frappe/frappe/issues/20714 . Furthermore it improves the idea of pending PR https://github.com/frappe/frappe/pull/20716
This PR #22316 doesn't fix the problem enough. For example, if a subject has several different sequences and one of them cannot be decoded, then the entire subject is lost.
3. Deleting an Email account also removes the link between the existing Doc "Communication" and the Email account. As a result, the existing Doc "Communication" becomes unattached to any Email account. method "def on_trash(self):" in frappe/email/doctype/email_account/email_account.py
Re-adding an Email account does NOT bind an existing Doc "Communication" with message_id = self.mail.get("Message-ID") You need to relink an existing "communication" document to a newly created email account when the "Message-ID" matches. Otherwise it is not possible to re-sync Email after adding an Email account that was previously deleted
4. If the IMAP folder name uses an encoding other than utf-8, then the folder name must be enclosed in quotes and encoded in IMAP utf-7

Closes https://github.com/frappe/frappe/issues/18168 & https://github.com/frappe/frappe/issues/19059
This PR was created on the basis of #20860, which for some reason was closed without implementation.

IgorA100 committed:
Update receive.py Fix: pull unseen mails via imap and decode subject

While refactoring the email logic from v13 to v14 a few things have been missed. Since then, only "unseen" mails could be pulled, see #19059 and #18168.

Furthermore decoding the subject was not working and sometimes only the first chunk has been encoded, see #20808 and #20716

IgorA100 committed:
Update email_account.py Fix: incorrect apostrophes for folder name

A folder setup as INBOX will be added hier into a string as '"INBOX"' instead of 'INBOX'. This will cause nested issues while updating the UIDs within check_imap_uidvalidity for docType "IMAP Folder" within receive.py.

IgorA100 committed:
Update utils.py Fix: helper method for proper decoding a sequence

Introduced new helpder method for decoding "subject" from emails, see #20808 and #20716

IgorA100 committed:
Update receive.py Removing my unnecessary comments

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 commented:
Everything works perfectly.
I didn't find any problems.

fnznhnzn commented:
maharshivpatel, if you found time to review this pr it would be greatly appreciated!:)

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 committed:
fix: Decoding algorithm changed

Changed decode_sequence

- "ascii" replace to "unicode_escape"
- None replace to "UTF-8"

IgorA100 committed:
test: add test for decode in subject header mail

Use "decode_sequence" in email/utils.py

IgorA100 commented:
@maharshivpatel 
Your opinion?

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 commented:
@maharshivpatel 
Please look at my changes and let me know your opinion.
I see you have already fixed one problem, but using a different method https://github.com/frappe/frappe/pull/23242 And now I have a conflict.

maharshivpatel commented on frappe/email/doctype/email_account/email_account.py:
```diff
@@ -578,7 +578,7 @@ def process_mail(messages, append_to=None):
 					if email_server.select_imap_folder(folder.folder_name):
 						email_server.settings["uid_validity"] = folder.uidvalidity
-						messages = email_server.get_messages(folder=f'"{folder.folder_name}"') or {}
+						messages = email_server.get_messages(folder=f'{folder.folder_name}') or {}
```
This won't work if folder name have space in it Like "My Folder"

IgorA100 replied:
Thanks to you, I discovered another problem with folders, if the folders are encoded other than UTF-8
I'm working on fixing this issue.

IgorA100 committed:
Merge branch 'develop' into patch-7

IgorA100 committed:
fix: Support other codings than utf-8 in the IMAP folder name

If the IMAP folder name is in an encoding other than utf-8, you must encode the folder and use quotes

IgorA100 committed:
feat: Add support encode & decode IMAP UTF-7

If the IMAP folder name is in an encoding other than utf-8, you must encode the folder and use quotes

IgorA100 committed:
fix: Cancel changes

It was my mistake.

IgorA100 commented:
@maharshivpatel 
I have made changes to my code.
Please check and approve!

IgorA100 committed:
fix: The algorithm for removing quote from a folder name has been changed

Based on https://github.com/frappe/frappe/pull/23368

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 committed:
fix: erroneous line removed

Erroneous line removed

IgorA100 commented:
Waiting ...

Akshaths12 commented:
Hello, I'm experiencing a similar issue with one of my customers where only unseen emails are updating despite changing the Email sync settings to 'ALL Emails.' The solutions mentioned in this ticket could be beneficial for my client too. Could you please provide an update on when this PR will be available for use?

IgorA100 commented:
Regarding ALL/unread changes have been made to:
- https://github.com/frappe/frappe/pull/23242
- https://github.com/maharshivpatel/frappe/commit/c2309969d570243010c44ad8cc75ed9b1e3e4b06
- https://github.com/maharshivpatel/frappe/commit/250dbf6a1caa5b0ab9bf89fe0413fc3a2b81ba06
Here we only fixed the type of the UID variable (in some cases it may affect the work)

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

Akshaths12 commented:
> Regarding ALL/unread changes have been made to:
> 
> * [fix: remove quotes from IMAP Folder Name #23242](https://github.com/frappe/frappe/pull/23242)
> * [maharshivpatel@c230996](https://github.com/maharshivpatel/frappe/commit/c2309969d570243010c44ad8cc75ed9b1e3e4b06)
> * [maharshivpatel@250dbf6](https://github.com/maharshivpatel/frappe/commit/250dbf6a1caa5b0ab9bf89fe0413fc3a2b81ba06)
>   Here we only fixed the type of the UID variable (in some cases it may affect the work)

Thank you!

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 commented:
@maharshivpatel
Have you forgotten about me?

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

IgorA100 commented:
Any problems?

IgorA100 committed:
Merge branch 'frappe:develop' into patch-7

ankush commented on frappe/email/receive.py:
```diff
@@ -150,8 +151,15 @@ def connect_pop(self):
 				raise
 
+	def folder_encode(self, folder):
+		# It is not yet known whether the folder name with quotes is passed to the method or not. 
+		# To select or get status a imap folder, you must send the name with quotes & encode for support in codings other than utf-8
+		if folder[0] == folder[-1] == '"':
+			folder = folder[1:-1]
+		return ImapUtf7.encode(f'"{folder}"')
```
Can you cite some docs for this? libraries seem to be defaulting on `"INBOX"` without extra quotes.

IgorA100 replied:
Thank you very much for your answer.
I've already lost all hope :(
1. A few lines below you can see the code that currently (before my changes) used quotes.
`res = self.imap.select(f'"{folder}"')`
2. https://datatracker.ietf.org/doc/html/rfc3501#section-5.1
> 1) Any character which is one of the atom-specials (see the Formal
> Syntax) will require that the mailbox name be represented as a
> quoted string or literal.
3. It seems like, theoretically, imaplib should insert quotes itself, but it doesn’t.
4. My changes have been tested and they work.

ankush committed:
Merge branch 'develop' into patch-7

ankush committed:
style: format

IgorA100 committed:
Removed unnecessary personal remark

IgorA100 commented:
Ping :)

IgorA100 committed:
Merge branch 'develop' into patch-7

kittiu commented:
This PR still have problem with at least windows-874 according to this https://github.com/frappe/frappe/pull/26001#issuecomment-2066822623

IgorA100 commented:
> This PR still have problem with at least windows-874 according

I'll make corrections soon.

kittiu commented:
> > This PR still have problem with at least windows-874 according
> 
> I'll make corrections soon.

Please be noted about this fix https://github.com/frappe/frappe/pull/26088

IgorA100 commented:
In progress
